packages/backend-platform/src/WebsocketService.ts

```diff
@@ -0,0 +1,1348 @@
+  handler: WebSocketService['getConnectionInfo'];
+};
+
+export type WebSocketServiceGetSubscriptionByChannelAction = {
+  type: `BackendWebSocketService:getSubscriptionByChannel`;
+  handler: WebSocketService['getSubscriptionByChannel'];
+};
+
+export type WebSocketServiceIsChannelSubscribedAction = {
+  type: `BackendWebSocketService:isChannelSubscribed`;
+  handler: WebSocketService['isChannelSubscribed'];
+};
+
+export type WebSocketServiceAddChannelCallbackAction = {
+  type: `BackendWebSocketService:addChannelCallback`;
+  handler: WebSocketService['addChannelCallback'];
+};
+
+export type WebSocketServiceRemoveChannelCallbackAction = {
+  type: `BackendWebSocketService:removeChannelCallback`;
+  handler: WebSocketService['removeChannelCallback'];
+};
+
+export type WebSocketServiceGetChannelCallbacksAction = {
+  type: `BackendWebSocketService:getChannelCallbacks`;
+  handler: WebSocketService['getChannelCallbacks'];
+};
+
+export type WebSocketServiceActions =
+  | WebSocketServiceInitAction
+  | WebSocketServiceConnectAction
+  | WebSocketServiceDisconnectAction
+  | WebSocketServiceSendMessageAction
+  | WebSocketServiceSendRequestAction
+  | WebSocketServiceGetConnectionInfoAction
+  | WebSocketServiceGetSubscriptionByChannelAction
+  | WebSocketServiceIsChannelSubscribedAction
+  | WebSocketServiceAddChannelCallbackAction
+  | WebSocketServiceRemoveChannelCallbackAction
+  | WebSocketServiceGetChannelCallbacksAction;
+
+export type WebSocketServiceAllowedActions = never;
+
+export type WebSocketServiceAllowedEvents = never;
+
+// Event types for WebSocket connection state changes
+export type WebSocketServiceConnectionStateChangedEvent = {
+  type: 'BackendWebSocketService:connectionStateChanged';
+  payload: [WebSocketConnectionInfo];
+};
+
+export type WebSocketServiceEvents =
+  WebSocketServiceConnectionStateChangedEvent;
+
+export type WebSocketServiceMessenger = RestrictedMessenger<
+  typeof SERVICE_NAME,
+  WebSocketServiceActions | WebSocketServiceAllowedActions,
+  WebSocketServiceEvents | WebSocketServiceAllowedEvents,
+  WebSocketServiceAllowedActions['type'],
+  WebSocketServiceAllowedEvents['type']
+>;
+
+/**
+ * WebSocket Service with automatic reconnection, session management and direct callback routing
+ *
+ * Real-Time Performance Optimizations:
+ * - Fast path message routing (zero allocations)
+ * - Production mode removes try-catch overhead
+ * - Optimized JSON parsing with fail-fast
+ * - Direct callback routing bypasses event emitters
+ * - Memory cleanup and resource management
+ *
+ * Mobile Integration:
+ * Mobile apps should handle lifecycle events (background/foreground) by:
+ * 1. Calling disconnect() when app goes to background
+ * 2. Calling connect() when app returns to foreground
+ * 3. Calling destroy() on app termination
+ */
+export class WebSocketService {
+  readonly #messenger: WebSocketServiceMessenger;
+
+  readonly #options: Required<WebSocketServiceOptions>;
+
+  #ws!: WebSocket;
```

Comment:
Why can we assume this is set?

Reply:
Because #ws is set only when we successfully established the connection. In case of error #ws is undefined.